jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/paths/PathCollection.java

```diff
@@ -0,0 +1,291 @@
+            }
+            catch (IOException e)
+            {
+                if (error == null)
+                    error = new RuntimeException("Unable to release some paths");
+                error.addSuppressed(e);
+            }
+        }
+        if (error != null)
+            throw error;
+    }
+
+    /**
+     * Resolve the first path that exists against the collection of paths.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @see Path#resolve(String)
+     */
+    public Path resolveFirstExisting(String other)
+    {
+        return resolveFirst(other, Files::exists);
+    }
+
+    /**
+     * Resolve the first path that matches the predicate.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @param pathPredicate the predicate of {@code Path} to evaluate for first valid hit.
+     */
+    public Path resolveFirst(String other, Predicate<Path> pathPredicate)
+    {
+        for (Path path : this)
+        {
+            Path dest = path.resolve(other);
+            if (dest.startsWith(path) && pathPredicate.test(dest))
+                return dest;
+        }
+        return null;
+    }
+
+    /**
+     * Resolve all paths across the path collection that matches the provided other.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @param pathPredicate the predicate of {@code Path} to evaluate for valid hit.
+     * @return the list of {@link Path} objects that satisfy the combination of {@code other} and {@code Predicate<Path>}
+     */
+    public List<Path> resolveAll(String other, Predicate<Path> pathPredicate)
+    {
+        List<Path> ret = new ArrayList<>();
+        for (Path path : this)
+        {
+            Path dest = path.resolve(other);
+            if (dest.startsWith(path) && pathPredicate.test(dest))
+                ret.add(dest);
+        }
+        return ret;
+    }
+
+    /**
+     * Find all files (up to 300 levels deep) in the path collection that match the provided {@link BiPredicate}
+     * of &lt;{@link Path}, {@link BasicFileAttributes}&gt;
+     *
+     * @param pathPredicate the predicate to evaluate across all paths in this collection
+     * @return the stream of hits for the find operation.
+     */
+    public Stream<Path> find(BiPredicate<Path, BasicFileAttributes> pathPredicate)
+    {
+        Stream<Path> ret = Stream.of();
+
+        for (Path path : this)
+        {
+            try
+            {
+                Stream<Path> stream = Files.find(path, 300, pathPredicate);
+                ret = Stream.concat(ret, stream);
+            }
+            catch (IOException e)
+            {
+                e.printStackTrace();
+            }
+        }
+        return ret;
+    }
```

Comment:
We need to have a `resolve(URI other)` method, as that will be the primary path from a request to a resource.

Reply:
Do we want to be able to resolve anything but `URI`s?

Reply:
I'm leery of using only URI as resolve against.  
As that's not the contract on `Path.resolve(*)` which is either String or Path, not URI.
Treating PathCollection.resolve() and Path.resolve() the same seems like it would just lower the number of bugs.

Reply:
The whole point of the changes to URI handling in jetty-12 is that we now have an unambiguous URI encoded pathInContext.   We can't give that as a string to a Path.resolve or PathCollection.resolve, because it will have encodings in it like %20 and %2F.  We can't decode to a string, as that introduces the ambiguities again as we lose information (was the / a / or a %2F encoding?).

Hence we need to resolve via something equivalent to `baseResoure.toURI().resolve(new URI(pathInContext))`